packages/gatsby/src/utils/graphql-typegen/ts-codegen.ts

```diff
@@ -0,0 +1,136 @@
+import * as fs from "fs-extra"
+import { join } from "path"
+import { slash } from "gatsby-core-utils"
+import { codegen } from "@graphql-codegen/core"
+import { Kind } from "graphql"
+import type { Types } from "@graphql-codegen/plugin-helpers"
+import type { TypeScriptPluginConfig } from "@graphql-codegen/typescript/config"
+import type { TypeScriptDocumentsPluginConfig } from "@graphql-codegen/typescript-operations/config"
+import { AnyAction, Store } from "redux"
+import { CodeFileLoader } from "@graphql-tools/code-file-loader"
+import { loadDocuments } from "@graphql-tools/load"
+import { IGatsbyState, IStateProgram } from "../../redux/types"
+import { filterTargetDefinitions, stabilizeSchema } from "./utils"
+
+const OUTPUT_PATH = `src/gatsby-types.d.ts`
+const NAMESPACE = `Queries`
+
+const DEFAULT_TYPESCRIPT_CONFIG: Readonly<TypeScriptPluginConfig> = {
+  avoidOptionals: true,
+  immutableTypes: true,
+  // TODO: Better maybeValue
+  maybeValue: `T | undefined`,
+  noExport: true,
+  enumsAsTypes: true,
+  scalars: {
+    Date: `string`,
+    JSON: `any`,
+  },
+  useTypeImports: true,
+}
```

Comment:
I would add a comment on why these defaults were chosen unless they are the default default ones from codegen itself.

```suggestion
const DEFAULT_TYPESCRIPT_CONFIG: Readonly<TypeScriptPluginConfig> = {
  avoidOptionals: true,
  immutableTypes: true,
  // TODO: Better maybeValue
  maybeValue: `T | undefined`,
  noExport: true,
  enumsAsTypes: true,
  scalars: {
    Date: `string`,
    JSON: `Record<string, unknown>`,
  },
  useTypeImports: true,
}
```